File: 3532-path-existence-queries-in-a-graph-i/3532-path-existence-queries-in-a-graph-i.cpp
class Solution {
public:
    vector<bool> pathExistenceQueries(int n, vector<int>& nums, int maxDiff, vector<vector<int>>& queries) {
        vector<int> root(n);

        for (int i = 0; i < n; i++) {
            root[i] = i;
        }

        int start = 0;
        for (int i = 1; i < n; i++) {
            if (nums[i] - nums[i - 1] <= maxDiff) {
                root[i] = start;
            } else {
                start = i;
            }
        }

        vector<bool> ans;
        for (auto &q : queries) {
            int u = q[0];
            int v = q[1];
            ans.push_back(root[u] == root[v]);
        }

        return ans;
    }
};

File: 3532-path-existence-queries-in-a-graph-i/3532-path-existence-queries-in-a-graph-i.py
class Solution:
    def pathExistenceQueries(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[bool]:
        root=[i for i in range(n)]
        start=0
        for i in range(1,n):
            if nums[i]-nums[i-1]<=maxDiff:
                root[i]=start
            else :
                start=i
        
        ans=[]
        for u,v in queries:
            ans.append(root[u]==root[v])

        return ans
        